SparkRunner: ValueIterator iterated twice during tuple encoding

Observing an error when running transformations via the SparkRunner.

Scio: 0.8.2
Beam 2.19.0
Spark: 2.4.4
Scala 2.12.10

The error below occurs identically whether running on the cluster or direct in local mode with `SparkRunner`.  Unfortunately the stacktrace doesn't surface the transform in the scio code that triggered the error, so I tried to summarize the transforms taking place below the stacktrace with some simplified code.

```
java.lang.RuntimeException: Exception while trying to `encode` an instance of Tuple2: Can't decode field _2
	at com.spotify.scio.coders.instances.PairCoder.onErrorMsg(ScalaCoders.scala:61)
	at com.spotify.scio.coders.instances.PairCoder.encode(ScalaCoders.scala:67)
	at com.spotify.scio.coders.instances.PairCoder.encode(ScalaCoders.scala:52)
	at org.apache.beam.sdk.coders.Coder.encode(Coder.java:136)
	at org.apache.beam.sdk.coders.NullableCoder.encode(NullableCoder.java:73)
	at org.apache.beam.sdk.coders.NullableCoder.encode(NullableCoder.java:63)
	at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply$mcV$sp(Coder.scala:221)
	at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
	at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
	at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:211)
	at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:221)
	at org.apache.beam.sdk.coders.Coder.encode(Coder.java:136)
	at org.apache.beam.sdk.coders.KvCoder.encode(KvCoder.java:71)
	at org.apache.beam.sdk.coders.KvCoder.encode(KvCoder.java:36)
	at org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.encode(WindowedValue.java:588)
	at org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.encode(WindowedValue.java:579)
	at org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.encode(WindowedValue.java:539)
	at org.apache.beam.runners.spark.coders.CoderHelpers.toByteArray(CoderHelpers.java:53)
	at org.apache.beam.runners.spark.translation.GroupNonMergingWindowsFunctions.lambda$groupByKeyAndWindow$c9b6f5c4$1(GroupNonMergingWindowsFunctions.java:86)
	at org.apache.beam.runners.spark.translation.GroupNonMergingWindowsFunctions.lambda$bringWindowToKey$0(GroupNonMergingWindowsFunctions.java:129)
	at org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators$6.transform(Iterators.java:785)
	at org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.TransformedIterator.next(TransformedIterator.java:47)
	at scala.collection.convert.Wrappers$JIteratorWrapper.next(Wrappers.scala:44)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:494)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:149)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:55)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:411)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at ### Coder materialization stack ###.(:0)
	at com.spotify.scio.coders.CoderMaterializer$.beamImpl(CoderMaterializer.scala:62)
	at com.spotify.scio.coders.CoderMaterializer$.beamImpl(CoderMaterializer.scala:68)
	at com.spotify.scio.coders.CoderMaterializer$.beamImpl(CoderMaterializer.scala:68)
	at com.spotify.scio.coders.CoderMaterializer$.beam(CoderMaterializer.scala:51)
	at com.spotify.scio.coders.CoderMaterializer$.beam(CoderMaterializer.scala:35)
	at com.spotify.scio.coders.CoderMaterializer$.kvCoder(CoderMaterializer.scala:95)
	at com.spotify.scio.values.SCollection.groupBy(SCollection.scala:499)
	at com.spotify.scio.values.SCollection.groupBy$(SCollection.scala:495)
	at com.spotify.scio.values.SCollectionImpl.groupBy(SCollection.scala:1362)
	at com.spotify.scio.values.PairSCollectionFunctions$$anonfun$asMultiMapSingletonSideInput$1.apply(PairSCollectionFunctions.scala:1194)
Caused by: java.lang.IllegalStateException: ValueIterator can't be iterated more than once,otherwise there could be data lost
	at org.apache.beam.runners.spark.translation.GroupNonMergingWindowsFunctions$GroupByKeyIterator$ValueIterator.iterator(GroupNonMergingWindowsFunctions.java:221)
	at scala.collection.convert.Wrappers$JIterableWrapper.iterator(Wrappers.scala:55)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at com.spotify.scio.coders.instances.SeqLikeCoder.encode(ScalaCoders.scala:158)
	at org.apache.beam.sdk.coders.Coder.encode(Coder.java:136)
	at org.apache.beam.sdk.coders.NullableCoder.encode(NullableCoder.java:73)
	at org.apache.beam.sdk.coders.NullableCoder.encode(NullableCoder.java:63)
	at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply$mcV$sp(Coder.scala:221)
	at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
	at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
	at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:211)
	at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:221)
	at com.spotify.scio.coders.instances.PairCoder$$anonfun$encode$4.apply$mcV$sp(ScalaCoders.scala:67)
	at com.spotify.scio.coders.instances.PairCoder$$anonfun$encode$4.apply(ScalaCoders.scala:67)
	at com.spotify.scio.coders.instances.PairCoder$$anonfun$encode$4.apply(ScalaCoders.scala:67)
	at com.spotify.scio.coders.instances.PairCoder.onErrorMsg(ScalaCoders.scala:55)
	at com.spotify.scio.coders.instances.PairCoder.encode(ScalaCoders.scala:67)
	at com.spotify.scio.coders.instances.PairCoder.encode(ScalaCoders.scala:52)
	at org.apache.beam.sdk.coders.Coder.encode(Coder.java:136)
	at org.apache.beam.sdk.coders.NullableCoder.encode(NullableCoder.java:73)
	at org.apache.beam.sdk.coders.NullableCoder.encode(NullableCoder.java:63)
	at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply$mcV$sp(Coder.scala:221)
	at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
	at com.spotify.scio.coders.WrappedBCoder$$anonfun$encode$1.apply(Coder.scala:221)
	at com.spotify.scio.coders.WrappedBCoder.catching(Coder.scala:211)
	at com.spotify.scio.coders.WrappedBCoder.encode(Coder.scala:221)
	at org.apache.beam.sdk.coders.Coder.encode(Coder.java:136)
	at org.apache.beam.sdk.coders.KvCoder.encode(KvCoder.java:71)
	at org.apache.beam.sdk.coders.KvCoder.encode(KvCoder.java:36)
	at org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.encode(WindowedValue.java:588)
	at org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.encode(WindowedValue.java:579)
	at org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.encode(WindowedValue.java:539)
	at org.apache.beam.runners.spark.coders.CoderHelpers.toByteArray(CoderHelpers.java:53)
	at org.apache.beam.runners.spark.translation.GroupNonMergingWindowsFunctions.lambda$groupByKeyAndWindow$c9b6f5c4$1(GroupNonMergingWindowsFunctions.java:86)
	at org.apache.beam.runners.spark.translation.GroupNonMergingWindowsFunctions.lambda$bringWindowToKey$0(GroupNonMergingWindowsFunctions.java:129)
	at org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators$6.transform(Iterators.java:785)
	at org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.TransformedIterator.next(TransformedIterator.java:47)
	at scala.collection.convert.Wrappers$JIteratorWrapper.next(Wrappers.scala:44)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:494)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:149)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:55)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:411)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at ### Coder materialization stack ###.(:0)
	at com.spotify.scio.coders.CoderMaterializer$.beamImpl(CoderMaterializer.scala:62)
	at com.spotify.scio.coders.CoderMaterializer$.beamImpl(CoderMaterializer.scala:68)
	at com.spotify.scio.coders.CoderMaterializer$.beamImpl(CoderMaterializer.scala:67)
	at com.spotify.scio.coders.CoderMaterializer$.beamImpl(CoderMaterializer.scala:68)
	at com.spotify.scio.coders.CoderMaterializer$.beam(CoderMaterializer.scala:51)
	at com.spotify.scio.coders.CoderMaterializer$.beam(CoderMaterializer.scala:35)
	at com.spotify.scio.coders.CoderMaterializer$.kvCoder(CoderMaterializer.scala:95)
	at com.spotify.scio.values.SCollection.groupBy(SCollection.scala:499)
	at com.spotify.scio.values.SCollection.groupBy$(SCollection.scala:495)
	at com.spotify.scio.values.SCollectionImpl.groupBy(SCollection.scala:1362)
```

```scala
val a = sc.avroFile[SpecificRecord](...)
val b = sc.avroFile[SpecificRecord](...)
val c = sc.avroFile[SpecificRecord](...)
val d = sc.avroFile[SpecificRecord](...)
val topN = 200
val minCountToTake = 2000

val aa = a
      .filter(_.stringField.startsWith("abc:"))
      .keyBy(_.stringField)
      .hashJoin(b.keyBy(_.stringField))
      .values
      .map(v => SomeCase(v._2.stringFieldA, v._1.stringFieldB, v._1.longField))

val ab = a
      .filter(!_.stringField.startsWith("abc"))
      .union(a)
      .map(tsbc => ((tsbc.stringFieldA, tsbc.stringFieldB), tsbc.longField))
      .sumByKey
      .map(summed => OtherCase(summed._1._1, summed._1._2, summed._2))
 
val ac =
    c.keyBy(_.idField)
      .join(ab.keyBy(_.idField))
      .values
      .map{joined =>
        val key = (
          joined._1.idField,
          joined._1.stringField
          joined._2.reportingCountry
        )
        (key, joined._2.chartsStreams)
      }.sumByKey
      .map(summed =>
        OtherCase(summed._1._1, summed._1._2, summed._1._3, summed._2))

    val ad = d.keyBy(_.idField)
      .join(ac.keyBy(_.id_field))
      .values
      .map(joined => OtherCase(
        joined._2.idField, joined._1,
        joined._2.stringField, Some(joined._2.LongField)))

    ad.map(r => (r.reportingRegion,
      FinalEntry(r.idField, r.stringField, r.optionalLongField.getOrElse(topN + 1))))
      .topByKey(topN)
      // using zip to give each recording an index as rank number
      // the range of index is 1 to topN
      .map(topped => {
        val ranked = topped._2.zip(Stream from 1)
        (topped._1, ranked.toList)
      })
      .flatMapValues(_.filter(_._2 > minCountToTake))
      .map(r => {
        val region = r._1
        val entry = r._2._1
        val rank = r._2._2
        OutputAvroType(
          entry.recordingGid, entry.stringField, region, entry.longField, rank)
      }).saveAsAvroFile[OutputAvroType](...)
  }


```

[Error in beam](https://github.com/apache/beam/blob/master/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupNonMergingWindowsFunctions.java#L222)
[Error in Scio](https://github.com/spotify/scio/blob/master/scio-core/src/main/scala/com/spotify/scio/coders/instances/ScalaCoders.scala#L59)

--- a/scio-core/src/main/scala/com/spotify/scio/coders/instances/ScalaCoders.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/coders/instances/ScalaCoders.scala
@@ -130,16 +130,6 @@ abstract private class BaseSeqLikeCoder[M[_], T](val elemCoder: BCoder[T])(
   // delegate methods for determinism and equality checks
   override def verifyDeterministic(): Unit = elemCoder.verifyDeterministic()
   override def consistentWithEquals(): Boolean = elemCoder.consistentWithEquals()
-  override def structuralValue(value: M[T]): AnyRef =
-    if (consistentWithEquals()) {
-      value.asInstanceOf[AnyRef]
-    } else {
-      val b = Seq.newBuilder[AnyRef]
-      val traversable = toSeq(value)
-      b.sizeHint(traversable.size)
-      traversable.foreach(v => b += elemCoder.structuralValue(v))
-      b.result()
-    }
 
   // delegate methods for byte size estimation
   override def isRegisterByteSizeObserverCheap(value: M[T]): Boolean = false
@@ -172,6 +162,17 @@ abstract private class SeqLikeCoder[M[_], T](bc: BCoder[T])(
     builder.result()
   }
 
+  override def structuralValue(value: M[T]): AnyRef =
+    if (consistentWithEquals()) {
+      value.asInstanceOf[AnyRef]
+    } else {
+      val b = Seq.newBuilder[AnyRef]
+      val traversable = ev(value)
+      b.sizeHint(traversable.size)
+      traversable.foreach(v => b += elemCoder.structuralValue(v))
+      b.result()
+    }
+
   override def toString: String = s"SeqLikeCoder($bc)"
 }
 
@@ -198,6 +199,12 @@ abstract private class BufferedSeqLikeCoder[M[_], T](bc: BCoder[T])(
       }
     }
     builder.result()
+  }
+
+  override def structuralValue(value: M[T]): AnyRef = {
+    val b = Seq.newBuilder[AnyRef]
+    ev(value).foreach(v => b += elemCoder.structuralValue(v))
+    b.result()
   }
 
   override def toString: String = s"BufferedSeqLikeCoder($bc)"

--- a/scio-core/src/main/scala/com/spotify/scio/coders/instances/ScalaCoders.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/coders/instances/ScalaCoders.scala
@@ -26,6 +26,8 @@ import org.apache.beam.sdk.coders.Coder.NonDeterministicException
 import org.apache.beam.sdk.coders.{Coder => BCoder, _}
 import org.apache.beam.sdk.util.CoderUtils
 import org.apache.beam.sdk.util.common.ElementByteSizeObserver
+import org.apache.beam.sdk.util.BufferedElementCountingOutputStream
+import org.apache.beam.sdk.util.VarInt
 
 import scala.collection.JavaConverters._
 import scala.reflect.ClassTag
@@ -151,27 +153,54 @@ abstract private class BaseSeqLikeCoder[M[_], T](val elemCoder: BCoder[T])(
 }
 
 abstract private class SeqLikeCoder[M[_], T](bc: BCoder[T])(
-  implicit toSeq: M[T] => TraversableOnce[T]
+  implicit ev: M[T] => TraversableOnce[T]
 ) extends BaseSeqLikeCoder[M, T](bc) {
-  protected val lc = VarIntCoder.of()
   override def encode(value: M[T], outStream: OutputStream): Unit = {
-    val traversable = toSeq(value)
-    lc.encode(traversable.size, outStream)
+    val traversable = ev(value)
+    VarInt.encode(traversable.size, outStream)
     traversable.foreach(bc.encode(_, outStream))
   }
+
   def decode(inStream: InputStream, builder: m.Builder[T, M[T]]): M[T] = {
-    val size = lc.decode(inStream)
+    val size = VarInt.decodeInt(inStream)
     builder.sizeHint(size)
     var i = 0
     while (i < size) {
       builder += bc.decode(inStream)
-      i = i + 1
+      i += 1
     }
     builder.result()
   }
 
-  override def toString: String =
-    s"SeqLikeCoder($bc)"
+  override def toString: String = s"SeqLikeCoder($bc)"
+}
+
+abstract private class BufferedSeqLikeCoder[M[_], T](bc: BCoder[T])(
+  implicit ev: M[T] => TraversableOnce[T]
+) extends BaseSeqLikeCoder[M, T](bc) {
+
+  override def encode(value: M[T], outStream: OutputStream): Unit = {
+    val buff = new BufferedElementCountingOutputStream(outStream)
+    ev(value).foreach { elem =>
+      buff.markElementStart()
+      bc.encode(elem, buff)
+    }
+    buff.finish()
+  }
+
+  def decode(inStream: InputStream, builder: m.Builder[T, M[T]]): M[T] = {
+    var count = VarInt.decodeLong(inStream);
+    while (count > 0L) {
+      builder += bc.decode(inStream)
+      count -= 1
+      if (count == 0L) {
+        count = VarInt.decodeLong(inStream);
+      }
+    }
+    builder.result()
+  }
+
+  override def toString: String = s"BufferedSeqLikeCoder($bc)"
 }
 
 private class OptionCoder[T](bc: BCoder[T]) extends SeqLikeCoder[Option, T](bc) {
@@ -198,14 +227,13 @@ private class ListCoder[T](bc: BCoder[T]) extends SeqLikeCoder[List, T](bc) {
   override def decode(inStream: InputStream): List[T] = decode(inStream, List.newBuilder[T])
 }
 
-// TODO: implement chunking
-private class TraversableOnceCoder[T](bc: BCoder[T]) extends SeqLikeCoder[TraversableOnce, T](bc) {
+private class TraversableOnceCoder[T](bc: BCoder[T])
+    extends BufferedSeqLikeCoder[TraversableOnce, T](bc) {
   override def decode(inStream: InputStream): TraversableOnce[T] =
     decode(inStream, Seq.newBuilder[T])
 }
 
-// TODO: implement chunking
-private class IterableCoder[T](bc: BCoder[T]) extends SeqLikeCoder[Iterable, T](bc) {
+private class IterableCoder[T](bc: BCoder[T]) extends BufferedSeqLikeCoder[Iterable, T](bc) {
   override def decode(inStream: InputStream): Iterable[T] =
     decode(inStream, Iterable.newBuilder[T])
 }
@@ -218,7 +246,7 @@ private class ArrayCoder[@specialized(Short, Int, Long, Float, Double, Boolean, 
   bc: BCoder[T]
 ) extends SeqLikeCoder[Array, T](bc) {
   override def decode(inStream: InputStream): Array[T] = {
-    val size = lc.decode(inStream)
+    val size = VarInt.decodeInt(inStream)
     val arr = new Array[T](size)
     var i = 0
     while (i < size) {

